File: frontend/src/services/marinha.js
// Nomes, cores e formatação da previsão marítima (Open-Meteo Marine,
// via backend — ver backend/clima/marinha.py).

// Nome em português de cada variável da API.
export const ROTULO_VARIAVEL = {
  wave_height: 'Altura de onda',
  wave_direction: 'Direção da onda',
  wave_period: 'Período da onda',
  wave_peak_period: 'Período de pico',
  wave_energy: 'Energia das ondas',
  wave_power: 'Potência das ondas',
  wind_wave_height: 'Onda de vento — altura',
  wind_wave_direction: 'Onda de vento — direção',
  wind_wave_period: 'Onda de vento — período',
  wind_wave_peak_period: 'Onda de vento — período de pico',
  swell_wave_height: 'Swell primário — altura',
  swell_wave_direction: 'Swell primário — direção',
  swell_wave_period: 'Swell primário — período',
  swell_wave_peak_period: 'Swell primário — período de pico',
  secondary_swell_wave_height: 'Swell secundário — altura',
  secondary_swell_wave_period: 'Swell secundário — período',
  secondary_swell_wave_direction: 'Swell secundário — direção',
  tertiary_swell_wave_height: 'Swell terciário — altura',
  tertiary_swell_wave_period: 'Swell terciário — período',
  tertiary_swell_wave_direction: 'Swell terciário — direção',
  sea_level_height_msl: 'Nível do mar',
  sea_surface_temperature: 'Temperatura da água',
  ocean_current_velocity: 'Velocidade da corrente',
  ocean_current_direction: 'Direção da corrente',
  invert_barometer_height: 'Efeito barômetro inverso',
}

// Nome curto pra cabeçalho de tabela.
export const ROTULO_CURTO = {
  wave_height: 'Onda',
  wave_direction: 'Dir. onda',
  wave_period: 'Período',
  wave_peak_period: 'Pico',
  wave_energy: 'Energia',
  wave_power: 'Potência',
  wind_wave_height: 'Vento alt.',
  wind_wave_direction: 'Vento dir.',
  wind_wave_period: 'Vento per.',
  wind_wave_peak_period: 'Vento pico',
  swell_wave_height: 'Swell alt.',
  swell_wave_direction: 'Swell dir.',
  swell_wave_period: 'Swell per.',
  swell_wave_peak_period: 'Swell pico',
  secondary_swell_wave_height: 'Swell 2 alt.',
  secondary_swell_wave_period: 'Swell 2 per.',
  secondary_swell_wave_direction: 'Swell 2 dir.',
  tertiary_swell_wave_height: 'Swell 3 alt.',
  tertiary_swell_wave_period: 'Swell 3 per.',
  tertiary_swell_wave_direction: 'Swell 3 dir.',
  sea_level_height_msl: 'Nível',
  sea_surface_temperature: 'Água',
  ocean_current_velocity: 'Corrente',
  ocean_current_direction: 'Dir. corrente',
  invert_barometer_height: 'Barôm. inv.',
}

// Cor fixa por modelo (tokens --modelo-N em theme.css). O best_match não
// entra no gráfico de comparação — ele é uma mistura dos outros.
export const COR_MODELO = {
  meteofrance_wave: 'var(--modelo-1)',
  meteofrance_currents: 'var(--modelo-2)',
  ecmwf_wam: 'var(--modelo-3)',
  ecmwf_wam025: 'var(--modelo-4)',
  ncep_gfswave025: 'var(--modelo-5)',
  dwd_gwam: 'var(--modelo-6)',
}

export const NOME_CURTO_MODELO = {
  best_match: 'Padrão',
  meteofrance_wave: 'MeteoFrance',
  meteofrance_currents: 'MF Currents',
  ecmwf_wam: 'ECMWF',
  ecmwf_wam025: 'ECMWF 0,25°',
  ncep_gfswave025: 'GFS',
  dwd_gwam: 'DWD',
}

// Escala de cores da potência das ondas (kW por metro de crista), do azul
// (fraca) ao magenta (muito forte) — usada nas barras do gráfico de energia.
// Energia e potência são calculadas no backend a partir da altura e do
// período (a Marine API não fornece esses dados).
export const ESCALA_POTENCIA = [
  { ate: 1, cor: '#3d6fc2' },
  { ate: 2, cor: '#4f9be0' },
  { ate: 4, cor: '#5cc6ee' },
  { ate: 6, cor: '#6fe6f2' },
  { ate: 8, cor: '#78f2d4' },
  { ate: 10, cor: '#7ef0a6' },
  { ate: 13, cor: '#7cec72' },
  { ate: 16, cor: '#a4ee58' },
  { ate: 20, cor: '#cdf24f' },
  { ate: 25, cor: '#f4f455' },
  { ate: 30, cor: '#f7d448' },
  { ate: 40, cor: '#f4ae3e' },
  { ate: 50, cor: '#ef8a33' },
  { ate: 65, cor: '#e75b27' },
  { ate: 80, cor: '#d93030' },
  { ate: 100, cor: '#b92a6a' },
  { ate: Infinity, cor: '#d93fd6' },
]

export function corDaPotencia(kwm) {
  if (kwm == null) return '#8fa3ad'
  return ESCALA_POTENCIA.find((faixa) => kwm < faixa.ate).cor
}

const DIRECAO_COM_ARTIGO = [
  'do Norte', 'de Nordeste', 'de Leste', 'de Sudeste', 'do Sul', 'de Sudoeste', 'de Oeste', 'de Noroeste',
]

// "do Sul", "de Sudeste"... pra nomear um swell pela direção de onde vem.
export function direcaoComArtigo(graus) {
  if (graus == null) return ''
  return DIRECAO_COM_ARTIGO[Math.round(graus / 45) % 8]
}

// Média de direções em graus (vetorial: a média de 350° e 10° é 0°).
export function mediaDirecoes(graus) {
  const validos = graus.filter((g) => g != null)
  if (!validos.length) return null
  const seno = validos.reduce((soma, g) => soma + Math.sin((g * Math.PI) / 180), 0)
  const cosseno = validos.reduce((soma, g) => soma + Math.cos((g * Math.PI) / 180), 0)
  return ((Math.atan2(seno, cosseno) * 180) / Math.PI + 360) % 360
}

export function ehDirecao(variavel) {
  return variavel.includes('direction')
}

const PONTOS_CARDEAIS = ['N', 'NE', 'L', 'SE', 'S', 'SO', 'O', 'NO']

export function direcaoTexto(graus) {
  if (graus == null) return '—'
  return PONTOS_CARDEAIS[Math.round(graus / 45) % 8]
}

// Onda: a API dá a direção DE ONDE vem (convenção meteorológica), então a
// seta aponta pra onde ela vai (+180°). Corrente: a API já dá PRA ONDE vai.
export function rotacaoSeta(variavel, graus) {
  if (graus == null) return null
  return variavel === 'ocean_current_direction' ? graus : (graus + 180) % 360
}

export function formatarValor(valor, unidade, casas = 1) {
  if (valor == null) return '—'
  const numero = Number(valor).toLocaleString('pt-BR', { minimumFractionDigits: 0, maximumFractionDigits: casas })
  if (!unidade) return numero
  return unidade === '°' || unidade === '°C' ? `${numero}${unidade}` : `${numero} ${unidade}`
}

export function formatarDataHora(dataHoraISO) {
  if (!dataHoraISO) return '—'
  return `${dataHoraISO.slice(8, 10)}/${dataHoraISO.slice(5, 7)} ${dataHoraISO.slice(11, 16)}`
}

export function formatarData(dataISO) {
  return dataISO ? `${dataISO.slice(8, 10)}/${dataISO.slice(5, 7)}` : '—'
}

export function formatarHora(dataHoraISO) {
  return dataHoraISO ? dataHoraISO.slice(11, 16) : '—'
}

// Índice da primeira hora >= agora (as horas vêm no fuso de São Paulo, sem
// offset — `new Date` lê no fuso do navegador, que no Brasil bate).
export function indiceAgora(tempos) {
  const agora = new Date()
  agora.setMinutes(0, 0, 0)
  const indice = tempos.findIndex((t) => new Date(t) >= agora)
  return indice === -1 ? 0 : indice
}

// Preamar/baixa-mar: máximos e mínimos locais da série de nível do mar
// (15 em 15 min). Janela de ±2 h pra não pegar oscilação pequena como maré.
export function extremosDeMare(tempos, niveis) {
  const janela = 8
  const extremos = []
  for (let i = janela; i < niveis.length - janela; i += 1) {
    const valor = niveis[i]
    if (valor == null) continue
    const vizinhos = niveis.slice(i - janela, i + janela + 1).filter((v) => v != null)
    const ehMaximo = vizinhos.every((v) => v <= valor) && niveis[i - 1] !== valor
    const ehMinimo = vizinhos.every((v) => v >= valor) && niveis[i - 1] !== valor
    if (ehMaximo) extremos.push({ tipo: 'alta', dataHora: tempos[i], nivel: valor })
    else if (ehMinimo) extremos.push({ tipo: 'baixa', dataHora: tempos[i], nivel: valor })
  }
  return extremos
}

File: backend/clima/marinha.py
"""Previsão marítima — Open-Meteo Marine Weather API (plano gratuito).

Usa tudo que a API oferece de graça e que tem dado na costa do RJ:

- `current`: as 23 variáveis, no modelo padrão (best_match).
- `hourly`: as 23 variáveis em cada modelo com cobertura no RJ (ver MODELOS).
- `daily`: as 11 variáveis diárias, em todos os modelos; o dia usa o
  modelo padrão e completa o que faltar com os outros (ver PRIORIDADE_DIARIA).
- `minutely_15`: nível do mar e correntes a cada 15 min.
- `era5_ocean`: reanálise histórica (últimos 92 dias, o máximo da API).

A Marine API não tem vento — ele vem da Forecast API da Open-Meteo, no
mesmo ponto (usado pela "Visão intermediária": terral/maral, rajadas).

A API também não tem energia nem potência das ondas: são calculadas aqui
(ver `energia_onda` e `potencia_onda`), pra cada componente (total, vagas
e swells) de todos os modelos.

Modelos que existem na API mas NÃO têm dado no RJ (testado): `dwd_ewam`
(só Europa) e `ncep_gfswave016` (só 52.5°N–15°S). Ficam de fora.
"""
import math
from concurrent.futures import ThreadPoolExecutor
from datetime import date

import requests

MARINE_URL = 'https://marine-api.open-meteo.com/v1/marine'
FORECAST_URL = 'https://api.open-meteo.com/v1/forecast'
TIMEOUT = 20
FUSO = 'America/Sao_Paulo'

VARIAVEIS_HORARIAS = [
    'wave_height', 'wave_direction', 'wave_period', 'wave_peak_period',
    'wind_wave_height', 'wind_wave_direction', 'wind_wave_period', 'wind_wave_peak_period',
    'swell_wave_height', 'swell_wave_direction', 'swell_wave_period', 'swell_wave_peak_period',
    'secondary_swell_wave_height', 'secondary_swell_wave_period', 'secondary_swell_wave_direction',
    'tertiary_swell_wave_height', 'tertiary_swell_wave_period', 'tertiary_swell_wave_direction',
    'sea_level_height_msl', 'sea_surface_temperature',
    'ocean_current_velocity', 'ocean_current_direction', 'invert_barometer_height',
]

VARIAVEIS_DIARIAS = [
    'wave_height_max', 'wave_direction_dominant', 'wave_period_max',
    'wind_wave_height_max', 'wind_wave_direction_dominant', 'wind_wave_period_max', 'wind_wave_peak_period_max',
    'swell_wave_height_max', 'swell_wave_direction_dominant', 'swell_wave_period_max', 'swell_wave_peak_period_max',
]

VARIAVEIS_15_MIN = ['sea_level_height_msl', 'ocean_current_velocity', 'ocean_current_direction']

VARIAVEIS_ERA5 = ['wave_height', 'wave_direction', 'wave_period']

# Ordem = ordem das cores no gráfico de comparação (não reordenar: a cor
# segue o modelo, nunca a posição na lista).
MODELOS = [
    {
        'id': 'meteofrance_wave', 'nome': 'MeteoFrance Wave', 'fornecedor': 'Serviço de meteorologia da França',
        'descricao': 'Calcula as ondas com bastante detalhe — é o que melhor enxerga a costa aqui.',
        'resolucao': '8 km', 'horizonte': '10 dias', 'atualizacao': 'a cada 12 horas',
    },
    {
        'id': 'meteofrance_currents', 'nome': 'MeteoFrance Currents', 'fornecedor': 'Serviço de meteorologia da França',
        'descricao': 'Não calcula ondas: cuida das correntes, da temperatura da água e do nível do mar (maré).',
        'resolucao': '8 km', 'horizonte': '10 dias', 'atualizacao': '1 vez por dia',
    },
    {
        'id': 'ecmwf_wam', 'nome': 'ECMWF WAM', 'fornecedor': 'Centro europeu de previsão do tempo',
        'descricao': 'Um dos modelos mais respeitados do mundo. Aqui dá só o básico da onda, mas prevê até 15 dias.',
        'resolucao': '9 km', 'horizonte': '15 dias', 'atualizacao': 'a cada 6 horas',
    },
    {
        'id': 'ecmwf_wam025', 'nome': 'ECMWF WAM 0,25°', 'fornecedor': 'Centro europeu de previsão do tempo',
        'descricao': 'Versão menos detalhada do anterior. Serve de segunda opinião pra conferir se os dois concordam.',
        'resolucao': '25 km', 'horizonte': '15 dias', 'atualizacao': 'a cada 6 horas',
    },
    {
        'id': 'ncep_gfswave025', 'nome': 'GFS Wave', 'fornecedor': 'Serviço de meteorologia dos EUA',
        'descricao': 'O que prevê mais longe (16 dias) e o que separa mais tipos de onda, incluindo 3 swells diferentes.',
        'resolucao': '25 km', 'horizonte': '16 dias', 'atualizacao': 'a cada 6 horas',
    },
    {
        'id': 'dwd_gwam', 'nome': 'DWD GWAM', 'fornecedor': 'Serviço de meteorologia da Alemanha',
        'descricao': 'Prevê só até 8 dias, mas é mais uma opinião independente pros primeiros dias.',
        'resolucao': '25 km', 'horizonte': '8 dias', 'atualizacao': 'a cada 12 horas',
    },
]

MODELO_ERA5 = {
    'id': 'era5_ocean', 'nome': 'ERA5 Ocean', 'fornecedor': 'Programa europeu Copernicus',
    'descricao': 'Não é previsão: é o registro do que o mar fez no passado. Alimenta o gráfico de Histórico.',
    'resolucao': '50 km', 'horizonte': 'desde 1940', 'atualizacao': '1 vez por dia',
}

MODELO_PADRAO = {
    'id': 'best_match', 'nome': 'Melhor combinação', 'fornecedor': 'Open-Meteo',
    'descricao': 'Não é um modelo à parte: a Open-Meteo escolhe sozinha o melhor dado pra cada informação. '
                 'É o que aparece em "Agora no mar" e nos primeiros dias.',
    'resolucao': '8 km', 'horizonte': '10 dias', 'atualizacao': 'automática',
}

# Ordem de preenchimento do card diário: modelo padrão, depois quem vai
# mais longe (GFS 16 dias, ECMWF 15 dias), depois o resto.
PRIORIDADE_DIARIA = ['best_match', 'ncep_gfswave025', 'ecmwf_wam', 'ecmwf_wam025', 'meteofrance_wave', 'dwd_gwam']

DIAS_SEMANA = ['Seg', 'Ter', 'Qua', 'Qui', 'Sex', 'Sáb', 'Dom']


DENSIDADE_AGUA_MAR = 1025  # kg/m³
GRAVIDADE = 9.81  # m/s²


def energia_onda(altura):
    """Energia das ondas em J/m² (densidade de energia): ρ·g·H²/16 — é a
    altura das barras no gráfico de energia."""
    if altura is None:
        return None
    return round(DENSIDADE_AGUA_MAR * GRAVIDADE * altura ** 2 / 16)


def potencia_onda(altura, periodo):
    """Potência das ondas em kW por metro de crista: ρ·g²·H²·T/(64π) —
    energia que a onda carrega por segundo; o período entra aqui, então
    ondas do mesmo tamanho mais espaçadas têm mais potência. É a cor das
    barras no gráfico de energia."""
    if altura is None or periodo is None:
        return None
    return round(DENSIDADE_AGUA_MAR * GRAVIDADE ** 2 * altura ** 2 * periodo / (64 * math.pi) / 1000, 1)


# Componentes que ganham energia/potência: total, vagas (ondas de vento) e
# os três swells. Prefixo das variáveis da API.
COMPONENTES_ONDA = ['wave', 'wind_wave', 'swell_wave', 'secondary_swell_wave', 'tertiary_swell_wave']


def _adicionar_energia(variaveis):
    for prefixo in COMPONENTES_ONDA:
        alturas = variaveis.get(f'{prefixo}_height')
        periodos = variaveis.get(f'{prefixo}_period')
        if not alturas:
            continue
        variaveis[f'{prefixo}_energy'] = [energia_onda(h) for h in alturas]
        if periodos:
            variaveis[f'{prefixo}_power'] = [potencia_onda(h, t) for h, t in zip(alturas, periodos)]


def _buscar(parametros, url=MARINE_URL):
    resposta = requests.get(url, params={**parametros, 'timezone': FUSO}, timeout=TIMEOUT)
    resposta.raise_for_status()
    return resposta.json()


def _tem_dado(lista):
    return any(valor is not None for valor in lista or [])


def _ultimo_indice_com_dado(listas):
    ultimo = -1
    for lista in listas:
        for indice in range(len(lista) - 1, ultimo, -1):
            if lista[indice] is not None:
                ultimo = indice
                break
    return ultimo


def _atual_e_15min(latitude, longitude):
    return _buscar({
        'latitude': latitude,
        'longitude': longitude,
        'current': ','.join(VARIAVEIS_HORARIAS),
        'minutely_15': ','.join(VARIAVEIS_15_MIN),
        'past_minutely_15': 24 * 4,  # último dia
        'forecast_minutely_15': 3 * 24 * 4,  # próximos 3 dias
    })


def _modelos(latitude, longitude):
    return _buscar({
        'latitude': latitude,
        'longitude': longitude,
        'hourly': ','.join(VARIAVEIS_HORARIAS),
        'daily': ','.join(VARIAVEIS_DIARIAS),
        'models': ','.join(['best_match'] + [m['id'] for m in MODELOS]),
        'forecast_days': 16,
    })


def _era5(latitude, longitude):
    return _buscar({
        'latitude': latitude,
        'longitude': longitude,
        'hourly': ','.join(VARIAVEIS_ERA5),
        'models': MODELO_ERA5['id'],
        'past_days': 92,
        'forecast_days': 1,
    })


def _vento(latitude, longitude):
    return _buscar({
        'latitude': latitude,
        'longitude': longitude,
        'hourly': 'wind_speed_10m,wind_direction_10m,wind_gusts_10m',
        'forecast_days': 10,
    }, url=FORECAST_URL)


def _normalizar_vento(dados):
    bruto = dados.get('hourly') or {}
    return {
        'tempo': bruto.get('time') or [],
        'velocidade': bruto.get('wind_speed_10m') or [],
        'direcao': bruto.get('wind_direction_10m') or [],
        'rajada': bruto.get('wind_gusts_10m') or [],
        'unidade': (dados.get('hourly_units') or {}).get('wind_speed_10m', 'km/h'),
    }


def _chave_modelo(variavel, modelo):
    # Com vários modelos na mesma chamada, cada variável vem com sufixo
    # (ex.: "wave_height_ecmwf_wam"); o best_match vem como "marine_best_match".
    sufixo = 'marine_best_match' if modelo == 'best_match' else modelo
    return f'{variavel}_{sufixo}'


def _unidades(*respostas):
    """{variavel: unidade} sem o sufixo de modelo — ex.: "wave_height": "m"."""
    sufixos = ['_marine_best_match'] + [f"_{m['id']}" for m in MODELOS + [MODELO_ERA5]]
    unidades = {}
    for dados in respostas:
        for grupo in ('current_units', 'hourly_units', 'daily_units', 'minutely_15_units'):
            for chave, unidade in (dados.get(grupo) or {}).items():
                if unidade in ('undefined', 'iso8601', 'seconds'):
                    continue
                base = next((chave[: -len(s)] for s in sufixos if chave.endswith(s)), chave)
                unidades.setdefault(base, unidade)
    return unidades


def _normalizar_horaria(dados):
    bruto = dados.get('hourly') or {}
    tempo = bruto.get('time') or []

    series = {}
    for modelo in ['best_match'] + [m['id'] for m in MODELOS]:
        variaveis = {}
        for variavel in VARIAVEIS_HORARIAS:
            valores = bruto.get(_chave_modelo(variavel, modelo))
            if _tem_dado(valores):
                variaveis[variavel] = valores
        _adicionar_energia(variaveis)
        if variaveis:
            series[modelo] = variaveis

    # Corta as horas do fim em que nenhum modelo tem mais nada.
    fim = _ultimo_indice_com_dado([v for vs in series.values() for v in vs.values()]) + 1
    return {
        'tempo': tempo[:fim],
        'series': {modelo: {var: valores[:fim] for var, valores in vs.items()} for modelo, vs in series.items()},
    }


def _normalizar_diaria(dados):
    bruto = dados.get('daily') or {}
    dias = []
    for indice, data_iso in enumerate(bruto.get('time') or []):
        dia = {'data': data_iso, 'diaSemana': DIAS_SEMANA[date.fromisoformat(data_iso).weekday()], 'fontes': {}}
        for variavel in VARIAVEIS_DIARIAS:
            valor, fonte = None, None
            for modelo in PRIORIDADE_DIARIA:
                lista = bruto.get(_chave_modelo(variavel, modelo)) or []
                if indice < len(lista) and lista[indice] is not None:
                    valor, fonte = lista[indice], modelo
                    break
            dia[variavel] = valor
            if fonte:
                dia['fontes'][variavel] = fonte
        dia['fonte'] = dia['fontes'].get('wave_height_max')
        if dia['fonte']:
            dias.append(dia)
    return dias


def _energia_maxima_por_dia(horaria):
    """{data: maior energia do dia}, na mesma ordem de prioridade dos cards
    diários: o modelo padrão primeiro, depois quem vai mais longe."""
    maximos = {}
    for modelo in PRIORIDADE_DIARIA:
        energias = horaria['series'].get(modelo, {}).get('wave_energy')
        if not energias:
            continue
        por_dia = {}
        for data_hora, energia in zip(horaria['tempo'], energias):
            if energia is not None:
                dia = data_hora[:10]
                por_dia[dia] = max(por_dia.get(dia, 0), energia)
        for dia, energia in por_dia.items():
            maximos.setdefault(dia, energia)
    return maximos


def _normalizar_15min(dados):
    bruto = dados.get('minutely_15') or {}
    return {
        'tempo': bruto.get('time') or [],
        **{variavel: bruto.get(variavel) or [] for variavel in VARIAVEIS_15_MIN},
    }


def _normalizar_era5(dados):
    bruto = dados.get('hourly') or {}
    tempo = bruto.get('time') or []
    variaveis = {v: bruto.get(v) or [] for v in VARIAVEIS_ERA5 if _tem_dado(bruto.get(v))}
    fim = _ultimo_indice_com_dado(list(variaveis.values())) + 1
    return {'tempo': tempo[:fim], **{v: valores[:fim] for v, valores in variaveis.items()}}


def buscar_previsao_maritima(latitude, longitude):
    with ThreadPoolExecutor(max_workers=4) as executor:
        futuro_atual = executor.submit(_atual_e_15min, latitude, longitude)
        futuro_modelos = executor.submit(_modelos, latitude, longitude)
        futuro_era5 = executor.submit(_era5, latitude, longitude)
        futuro_vento = executor.submit(_vento, latitude, longitude)
        atual_e_15min = futuro_atual.result()
        modelos = futuro_modelos.result()
        # Histórico e vento são opcionais: se falharem, o resto continua.
        try:
            era5 = futuro_era5.result()
        except requests.RequestException:
            era5 = {}
        try:
            vento = _normalizar_vento(futuro_vento.result())
        except requests.RequestException:
            vento = None

    horaria = _normalizar_horaria(modelos)
    atual = {k: v for k, v in (atual_e_15min.get('current') or {}).items() if k != 'interval'}
    atual['wave_energy'] = energia_onda(atual.get('wave_height'))
    atual['wave_power'] = potencia_onda(atual.get('wave_height'), atual.get('wave_period'))

    diaria = _normalizar_diaria(modelos)
    energia_por_dia = _energia_maxima_por_dia(horaria)
    for dia in diaria:
        dia['wave_energy_max'] = energia_por_dia.get(dia['data'])

    metadados = []
    for modelo in [MODELO_PADRAO] + MODELOS:
        variaveis = horaria['series'].get(modelo['id'], {})
        if not variaveis:
            continue
        horas = max(sum(1 for valor in valores if valor is not None) for valores in variaveis.values())
        metadados.append({**modelo, 'variaveis': list(variaveis), 'horasComDado': horas})

    historico = _normalizar_era5(era5)
    if historico['tempo']:
        metadados.append({
            **MODELO_ERA5,
            'variaveis': [v for v in VARIAVEIS_ERA5 if v in historico],
            'horasComDado': len(historico['tempo']),
        })

    return {
        'grade': {'latitude': modelos.get('latitude'), 'longitude': modelos.get('longitude')},
        'unidades': {
            **_unidades(atual_e_15min, modelos, era5),
            **{f'{p}_energy': 'J/m²' for p in COMPONENTES_ONDA},
            **{f'{p}_power': 'kW/m' for p in COMPONENTES_ONDA},
            'wave_energy_max': 'J/m²',
        },
        'atual': atual,
        'diaria': diaria,
        'horaria': horaria,
        'quinzeMinutos': _normalizar_15min(atual_e_15min),
        'historico': historico,
        'vento': vento,
        'modelos': metadados,
    }
